test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Parsers/ODataPathParserTests.cs

```diff
@@ -36,30 +36,72 @@ public void CallbackReturnsBatchReferenceSegment()
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: false, "multiple(123;321)", "multiple", "123;321");
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: false, "set()", "set", string.Empty);
+
+            // Failure cases
+            Action emptyString =
+                () =>
+                    ODataPathParser.ExtractSegmentIdentifierAndParenthesisExpression(
+                        enableKeyAsSegment: false,
+                        string.Empty,
+                        out actualIdentifier,
+                        out queryPortion);
             emptyString.Throws<ODataUnrecognizedPathException>(ErrorStrings.RequestUriProcessor_EmptySegmentInRequestUrl);
 
-            Action noIdentifier = () => ODataPathParser.ExtractSegmentIdentifierAndParenthesisExpression("()", out actualIdentifier, out queryPortion);
+            Action noIdentifier =
+                () =>
+                    ODataPathParser.ExtractSegmentIdentifierAndParenthesisExpression(
+                        enableKeyAsSegment: false,
+                        "()",
+                        out actualIdentifier,
+                        out queryPortion);
             noIdentifier.Throws<ODataUnrecognizedPathException>(ErrorStrings.RequestUriProcessor_EmptySegmentInRequestUrl);
 
-            Action noEndParen = () => ODataPathParser.ExtractSegmentIdentifierAndParenthesisExpression("foo(", out actualIdentifier, out queryPortion);
+            Action noEndParen =
+                () =>
+                    ODataPathParser.ExtractSegmentIdentifierAndParenthesisExpression(
+                        enableKeyAsSegment: false,
+                        "foo(",
+                        out actualIdentifier,
+                        out queryPortion);
             noEndParen.Throws<ODataException>(ErrorStrings.RequestUriProcessor_SyntaxError);
         }
 
+        [Fact]
+        public void RequestUriProcessorExtractSegmentIdentifierTest_KeyAsSegment()
+        {
+            string actualIdentifier;
+            string queryPortion;
+
+            // Success cases
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: true, "blah", "blah", null);
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: true, "multiple words", "multiple words", null);
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: true, "multiple(123)", "multiple", "123");
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: true, "multiple(123;321)", "multiple", "123;321");
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: true, "set()", "set", string.Empty);
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: true, "()", "()", null);
+            ExtractSegmentIdentifierAndParenthesisExpression(enableKeyAsSegment: true, "foo(", "foo(", null);

```

Comment:
Could you add a test case for a scenario that does not contain an `(` but ends with a `)`, like `blah)`.

Also, is it a good idea to support different syntaxes depending on whether key as segment is enabled?